[SYCL] Make Windows DLLs proxy loader optional

ph0b opened the pull request:
Customers who don't keep static refs to sycl objects don't specifically need this proxy loader workaround and may want to avoid having to distribute one more DLL.
This change keeps default behavior unchanged, the proxy loader stays enabled by default on Windows.

ph0b commented:
@cperkinsintel can you review ?

cperkinsintel reviewed (approved):
LGTM

ph0b commented:
> Is there really a demand for this not having to distribute an extra dll?
Yes. 

Blender has been shipping for a year to end-users with a SYCL runtime from prior the inclusion of this proxy loader, and no issue has been reported, so it still sounds OK.

Even if it was not OK, this PR is still one step in the right direction to trim down the number of DLLs to ship, and allow for easy testing on this issue.

We can imagine a next step in which ISVs will be able to compile the runtime with static / statically linked plugins. ISVs need to know the platforms they want to support at compile time anyway, the dynamic aspect of the plugin mechanism is of no value to them (Example: if you add nvidia plugin after an application is built, it will not be enough to enable nvidia support, as it will lack PTX).

cperkinsintel commented:
> Blender has been shipping for a year to end-users with a SYCL runtime from prior the inclusion of this proxy loader, and no issue has been reported, so it still sounds OK.

That's not a fair comparison. Before we introduced the proxy loader we just skipped all the shutdown procs on Windows.  So piTearDown was simply not called. Now it is. And if the plug-ins need to be called from inside piTearDown, you'll get a crash if Windows has unloaded them.

ph0b commented:
Interesting, so basically, this PR should skip the shutdown procs too? Are plugins currently called from piTearDown? Wouldn't there be a safe way to avoid crashing in that case without the proxy, by checking another way if it's still loaded?

cperkinsintel commented:
If you skip the piTearDown then you are at risk of leaving a device handle open, making the device unavailable to other apps. Though in practice, I don't think this actually happens, or at least it used to not.  There have been a lot of changes with the universal runtime, so I don't know. 

> Wouldn't there be a safe way to avoid crashing in that case without the proxy, by checking another way if it's still loaded?
The DLL itself gets notice that it is being destroyed, but the trick is communicating that back to the caller and then adding the checks on the call side. This is non-trivial and would have to extend beyond merely a check in piTearDown, so it might be an extensive change to the whole PI.  Then it becomes the question of the juice worth the squeeze? The proxy loader is a simpler solution and avoids the problem at a trivial cost.

ph0b commented:
Why each plugin can't deal with its own teardown? I don't get why it's the runtime responsibility to deal with it.

From your angle, as you worked on it, I understand the proxy loader is a clean solution, but from ISVs trying to adopt sycl for the first time, it's a novelty, they haven't seen anything similar in other frameworks, so from a first glance, it looks hacky and clunky, and it's not obvious how many libraries they need to redistribute (and it's not the same set across OSes), so while an improvement may not be as easy, it'd still be welcome.

ph0b commented:
I've rebased and added what was needed to avoid calling shutdown() in case of process exit if there is no proxy loader.

This is what https://learn.microsoft.com/en-us/windows/win32/dlls/dllmain recommends to do that in all cases:
> If the process is terminating (the lpvReserved parameter is non-NULL), all threads in the process except the current thread either have exited already or have been explicitly terminated by a call to the [ExitProcess](https://learn.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-exitprocess) function, which might leave some process resources such as heaps in an **inconsistent state**. In this case, it is **not safe for the DLL to clean up the resources**. Instead, the DLL should allow the operating system to reclaim the memory.

This paragraph from the documentation makes me again doubt the actual need for the proxy loader solution on windows, the potential issues you mention if they do exist shouldn't need a global hammer like this.
